src/Corax/Queries/SortingMatch.cs

```diff
@@ -135,15 +142,34 @@ public int Fill(Span<long> matches)
 
             Unsafe.SkipInit(out key);
             return false;
-        }
+        }        
 
         [SkipLocalsInit]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private int Fill<W>(Span<long> matches) where W : struct
-        {
-            int take = _take <= 0 ? matches.Length : _take;
+        {             
+            // Important: If you are going to request a massive take like 20K you need to pass at least a 20K size buffer to work with.
+            //            The rationale for such behavior is that sorting has to find among the candidates the order between elements,
+            //            and it can't do so without checking every single element found. If you fail to do so, your results may not be
+            //            correct. 
+            Debug.Assert(_take <= matches.Length);
+
+            Span<MatchComparer<TComparer, W>.Item> matchesKeys;
+            var matchesKeysHolder = MatchComparer<TComparer, W>.MatchKeysHolder;
+            if (matchesKeysHolder != null && matchesKeysHolder.Length > matches.Length)
+                matchesKeys = matchesKeysHolder.AsSpan(0, matches.Length);
+            else
+                matchesKeys = new MatchComparer<TComparer, W>.Item[matches.Length].AsSpan();
```

Comment:
PowerOfTwo here?
Or use a ArrayPool?

Reply:
I would have choose ArrayPool but you said to use a threadstatic, which is fast. I have no opinion on what it is best yet.

Reply:
The concern with a thread static is that we'll have that buffer stuck for quite some time. ArrayPool might be easier / better. IIRC, they use thread static internally to ensure this. Might want to use a local one, instead of the shared?

Reply:
If we go `ArrayPool` I believe we need a query array pool that should have high traffic anyways.